web/server.go

```diff
@@ -0,0 +1,69 @@
+package web
+
+import (
+	"context"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+// Server will create a http.Server from the Go standard library
+type Server struct {
+	Addr    string        // TCP address to listen on, ":http" if empty
+	Handler http.Handler  // handler to invoke, http.DefaultServeMux if nil
+	Timeout time.Duration // Timeout to wait for shutdown
+}
+
+// ListenAndServe mirrors the function from the Go standard library
+// ListenAndServe always returns a non-nil error
+func (s *Server) ListenAndServe() error {
+	server := http.Server{
+		Addr:    s.Addr,
+		Handler: s.Handler,
+	}
+
+	var grErr error
+	done := make(chan struct{})
+	// Start the listener.
+	go func() {
+		grErr = server.ListenAndServe()
+		close(done)
+	}()
+
+	// Listen for an interrupt signal from the OS. Use a buffered
+	// channel because of how the signal package is implemented.
+	osSignals := make(chan os.Signal, 1)
+	signal.Notify(osSignals, os.Interrupt, syscall.SIGTERM)
+
+	var sdErr error
+	// If server.ListenAndServe() cannot startup due to errors such as "port in use",
+	// it will return error and get stuck waiting for osSignals.
+	// This is not ideal because if a server doesn't start, we want to log the error and exit.
+	// Unfortunately, we can't select on a waitgroup.
+	// The done channel and select statement is used to handle the above case.
+	select {
+	case <-done:
+	case <-osSignals:
+		ctx, cancel := context.WithTimeout(context.Background(), s.Timeout)
+		defer cancel()
+
+		// Attempt the graceful shutdown by closing the listener and
+		// completing all inflight requests.
+		if err := server.Shutdown(ctx); err != nil {
+			sdErr = err
```

Comment:
I think don't need to assign the err here, since we are "handling" it by killing it hard?

Reply:
make sense

Reply:
What if `Shutdown` error-ed and `Close` was successful?

Reply:
It means it is successful, correct? The errors are handled and should not bubble up?

Reply:
Ok, we moved from the paradigm of `err == nil` means no error by stating upfront that this method returns a non-nil error response. Which was our understanding when we changed the method signature from returning `string` to returning `error`. So the thinking has been shifted to "which error should we return here?".

By this thinking, it seems better to return something one level deeper than what `ListenAndServe` returns by returning the error for `Shutdown` if `Shutdown` has error but `Close` does not.

Reply:
I see, okay that makes sense.

Reply:
so just to be clear, 
1. if startup fails, returns startup err (port in use etc.)
2. if gracefully shutdown, return startup err (which is server closed)
3. if ungracefully shutdown, return ungraceful err

Reply:
To disambiguate, `startup err` == `ListenAndServe error`.

Yes

Reply:
Can `server.Close()` fail? What does it return then?

Reply:
Reading the [documentation](https://golang.org/pkg/net/http/#Server.Close), it says the following

> Close returns any error returned from closing the Server's underlying Listener(s).